pkg/controllers/awsloadbalancercontroller/deployment_test.go

```diff
@@ -0,0 +1,386 @@
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            fmt.Sprintf("aws-load-balancer-operator-%s", b.name),
+			Namespace:       "test-namespace",
+			OwnerReferences: b.ownerReference,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					appInstanceName: b.name,
+					appLabelName:    appName,
+				},
+			},
+			Replicas: b.replicas,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						appInstanceName: b.name,
+						appLabelName:    appName,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers:         b.containers,
+					ServiceAccountName: b.serviceAccount,
+				},
+			},
+		},
+	}
+	if b.version != "" {
+		d.ResourceVersion = b.version
+	} else {
+		d.ResourceVersion = "1"
+	}
+	return d
+}
+
+func testContainer(name, image string, args ...string) corev1.Container {
+	return corev1.Container{
+		Name:  name,
+		Image: image,
+		Args:  args,
+	}
+}
+
+func TestUpdateDeployment(t *testing.T) {
+	for _, tc := range []struct {
+		name               string
+		existingDeployment *appsv1.Deployment
+		desiredDeployment  *appsv1.Deployment
+		expectedDeployment *appsv1.Deployment
+	}{
+		{
+			name: "image changed",
+			existingDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v1"),
+			).build(),
+			desiredDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v2"),
+			).build(),
+			expectedDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v2"),
+			).build(),
+		},
+		{
+			name: "replicas changed from value",
+			existingDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v1"),
+			).withReplicas(1).build(),
+			desiredDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v1"),
+			).withReplicas(2).build(),
+			expectedDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v1"),
+			).withReplicas(2).build(),
+		},
+		{
+			name: "replicas changed from nil",
+			existingDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v1"),
+			).build(),
+			desiredDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v1"),
+			).withReplicas(1).build(),
+			expectedDeployment: testDeployment("operator", "test-namespace", "test-sa").withContainers(
+				testContainer("controller", "controller:v1"),
+			).withReplicas(1).build(),
+		},
+		{
+			name: "container args changed",
```

Comment:
Can you please add a test case when an unsolicited container was added or desired removed?

Reply:
I'm actually allowing the addition of containers because there could be something else which adds a sidecar and I don't want the operator to stomp that if the desired containers are present.

Reply:
Will add tests for that however.

Reply:
Scratch the above as well, I am resetting the containers as well. Looks like I forgot how I implemented everything.